frontend/src/components/pages/cabins/booking/Steps/CabinSelection.tsx

```diff
@@ -0,0 +1,205 @@
+import { useQuery } from "@apollo/client";
+import { KeyboardArrowRight } from "@mui/icons-material";
+import { Button, Checkbox, Divider, FormControlLabel, Paper, Stack, Tooltip, Typography } from "@mui/material";
+import dayjs from "dayjs";
+import React from "react";
+
+import { Calendar } from "@/components/Calendar";
+import { BookingSemesterDocument, CabinFragment } from "@/generated/graphql";
+import { useOverlappingBookings } from "@/hooks/cabins/useBookedDates";
+import { useResponsive } from "@/hooks/useResponsive";
+
+import { Stepper } from "./Stepper";
+
+type Props = {
+  allCabins: CabinFragment[];
+  chosenCabins: CabinFragment[];
+  onCabinsChange: (cabins: CabinFragment[]) => void;
+  onDateChange: (date: dayjs.Dayjs) => void;
+  startDate: dayjs.Dayjs | undefined;
+  endDate: dayjs.Dayjs | undefined;
+  onNext: () => void;
+};
+
+type Validation =
+  | {
+      valid: false;
+      error: string;
+    }
+  | { valid: true; error?: never };
+
+/**
+ * One of the steps in the cabins/book page. In this step the user chooses a cabin and the check-in and check-out dates.
+ */
+export const CheckInOut: React.FC<React.PropsWithChildren<Props>> = ({
+  allCabins,
+  chosenCabins,
+  onCabinsChange,
+  onDateChange,
+  startDate,
+  endDate,
+  onNext,
+}) => {
+  const { isInActiveBookingSemester } = useIsInActiveBookingSemester();
+  const { hasOverlapWithOtherBookings, isDateWithOverlappingBooking } = useOverlappingBookings();
+  const isMobile = useResponsive({ query: "down", key: "md" });
+
+  /**
+   * Validate that at least one cabin has been chosen, and that the
+   * selected dates are valid, i.e. not overlapping with disabled dates.
+   */
+  function validateCabinAndDateSelection(
+    chosenCabins: CabinFragment[],
+    dates: { start: dayjs.Dayjs | undefined; end: dayjs.Dayjs | undefined }
+  ): Validation {
+    const { start, end } = dates;
+
+    if (chosenCabins.length === 0) {
+      return { valid: false, error: "Du må velge en hytte" };
+    }
+
+    // The user needs to enter a check-in date
+    if (start === undefined) {
+      return { valid: false, error: "Du må velge en dato for innsjekk" };
+    }
+    // The user needs to enter a check-out date
+    if (end === undefined) {
+      return { valid: false, error: "Du må velge en dato for utsjekk" };
+    }
+
+    if (hasOverlapWithOtherBookings({ start, end }, chosenCabins)) {
+      return { valid: false, error: "Denne datoen er allerede booket" };
+    }
+
+    return { valid: true };
+  }
+
+  const validation = validateCabinAndDateSelection(chosenCabins, { start: startDate, end: endDate });
+
+  /**
+   * Dates are disabled if:
+   * 1. No cabin is selected
+   * 2. The date is not in the active booking semester
+   * 3. The date has an overlapping booking
+   * 4. The date is before today
```

Comment:
You don't need to add this, might even be a nice first problem for new members next semester, but I know it has been requested that single dates be removed from the calendar. '(Its not possible to book from 1. january to 1.january)

Reply:
Great suggestion, I know this has been requested. I'll leave it as an exercise 😉